gamemodes/terrortown/entities/entities/ttt_smoke.lua

```diff
@@ -0,0 +1,428 @@
+		return value
+	end
+end
+
+---
+-- @realm shared
+function ENT:Think()
+	-- TODO: figure out if this is relevant
+	if ( CLIENT ) then
+		local physobj = self:GetPhysicsObject()
+
+		if ( IsValid( physobj ) ) then
+			physobj:SetPos( self:GetPos() )
+			physobj:SetAngles( self:GetAngles() )
+		end
+	end
+
+	if self:GetExtinguishFires() then
+		game_effects.ExtinguishInRadius(self:GetPos(), self:GetRadius())
+	end
+
+	if self:GetAlive() and CurTime() > (self:GetCreateTime() + self:GetLifeTime()) then
+		self:Disipate()
+	end
+	if SERVER and not self:GetAlive() and CurTime() > (self:GetDeactivateTime() + self:GetFadeTime()) then
+		self:Remove()
+	end
+end
+
+---
+-- Create a color from the smoke grenade.
+-- @todo this is called inside game_effects.SpawnSmoke which is SERVER only, this does not seem correct.
+-- @realm client
+function ENT:GenerateSmokeColor()
+	local RandomWrap = self.RandomWrap
+	local color_override = nil
+
+	if self:GetRandomColor() then
+		local colors_available = table.GetKeys(self.SmokeColors)
+		local color_dex = math.max(1, math.ceil(RandomWrap(0, #colors_available)))
+		local color_name = colors_available[color_dex]
+		print("picked color", color_name, "or", color_dex, "out of", #colors_available)
+		color_override = self.SmokeColors[ color_name ]
+	end
+
+	local mix = math.Clamp(RandomWrap(
+		(self:GetColorBase() - self:GetColorVariance()) / 255,
+		(self:GetColorBase() + self:GetColorVariance()) / 255
+	), 0, 1)
+
+	if color_override ~= nil then
+		local h, s = ColorToHSV( color_override )
+		color_override = HSVToColor(h, s, mix)
+	else
+		color_override = HSVToColor(0, 0, mix)
+		-- legacy:
+		-- local unmix = math.Round(mix*255)
+		-- Color(unmix, unmix, unmix)
+	end
+
+	return color_override
+end
+
+---
+-- Gets a point within a sphere that is suitably random, and uses our PRNG (SharedRandom).
+-- @todo move to a library or util, idk
+-- @realm shared
+function ENT:PointInSphere()
+	local RandomWrap = self.RandomWrap
+
+	local u = RandomWrap(0, 1)
+	local v = RandomWrap(0, 1)
+	local theta = u * 2.0 * math.pi
+	local phi = math.acos(2.0 * v - 1.0)
+	local r = math.pow(RandomWrap(0, 1), 1 / 2)
+	local sinTheta = math.sin(theta)
+	local cosTheta = math.cos(theta)
+	local sinPhi = math.sin(phi)
+	local cosPhi = math.cos(phi)
+	local x = r * sinPhi * cosTheta
+	local y = r * sinPhi * sinTheta
+	local z = r * cosPhi
+	return Vector(x, y, z)
+end
+
+---
+-- Initiates the smoke effect and spawns all the particles, by reading all the data on this entity.
+-- @realm client
+function ENT:ClientEffect()
+	local RandomWrap = self.RandomWrap
+	self.Emitter = ParticleEmitter(self:GetPos(), false)
+	self.Emitter:SetPos(self:GetPos())
+	local c = self:GetSmokeColor():ToColor()
+	local diameter = self:GetRadius() * 2
+	print("particles for:", self:GetRadius(), "d:", diameter, self:GetStartSizeBase(), "+/-",  self:GetStartSizeVariance(), "legend:")
+	print("particle:\tstart_size\tmax_range\toffset_ratio\toffset_dimension")
```

Comment:
remove prints and maybe put them in a nice order that it looks nice. In general I try to put the locals on top, add an empty line and then do the other suff